drivers/platform/mbed/mbed_gpio.cpp

```diff
@@ -0,0 +1,308 @@
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t mbed_gpio_get_optional(struct gpio_desc **desc,
+			       const struct gpio_init_param *param)
+{
+	if (param == NULL) {
+		*desc = NULL;
+		return SUCCESS;
+	}
+
+	return mbed_gpio_get(desc, param);
+}
+
+/**
+ * @brief Free the resources allocated by mbed_gpio_get().
+ * @param desc[in] - The GPIO descriptor.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t mbed_gpio_remove(struct gpio_desc *desc)
+{
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	/* Free the gpio object */
+	if (((struct mbed_gpio_desc *)(desc->extra))->gpio_pin)
+		free(((struct mbed_gpio_desc *)(desc->extra))->gpio_pin);
+
+	/* Free the gpio descriptor objects */
+	free(desc->extra);
+	free(desc);
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Enable the input direction of the specified GPIO.
+ * @param desc[in] - The GPIO descriptor.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ * @note does not support reconfiguration of already set pin direction
+ */
+int32_t mbed_gpio_direction_input(struct gpio_desc *desc)
+{
+	DigitalIn *gpio_input;				// pointer to gpio input object
+	struct mbed_gpio_desc
+		*gpio_desc_extra; 	// pointer to gpio desc extra parameters
+
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	gpio_desc_extra = (struct mbed_gpio_desc *)(desc->extra);
+	if (gpio_desc_extra->gpio_pin == NULL) {
+		/* Configure and instantiate GPIO pin as input */
+		gpio_input = new DigitalIn((PinName)desc->number);
+		if (!gpio_input)
+			return -ENOMEM;
+
+		gpio_desc_extra->gpio_pin = (struct mbed_gpio_desc *)gpio_input;
+		gpio_desc_extra->direction = GPIO_IN;
+
+		/* Set the gpio pin mode */
+		gpio_input->mode((PinMode)((struct mbed_gpio_desc *)desc->extra)->pin_mode);
+
+		return SUCCESS;
+	}
+
+	return -EINVAL;
+}
+
+/**
+ * @brief Enable the output direction of the specified GPIO.
+ * @param desc[in] - The GPIO descriptor.
+ * @param value[in] - The GPIO value.
+ *                Example: GPIO_HIGH
+ *                         GPIO_LOW
+ * @return SUCCESS in case of success, negative error code otherwise.
+ * @note does not support reconfiguration of already set pin direction
+ */
+int32_t mbed_gpio_direction_output(struct gpio_desc *desc, uint8_t value)
+{
+	DigitalOut *gpio_output;    	// pointer to gpio output object
+	struct mbed_gpio_desc
+		*gpio_desc_extra;   // pointer to gpio desc extra parameters
+
+	if (!desc|| !desc->extra)
+		return -EINVAL;
+
+	gpio_desc_extra = (struct mbed_gpio_desc *)(desc->extra);
+	if (gpio_desc_extra->gpio_pin == NULL) {
+		/* Configure and instantiate GPIO pin as output */
+		gpio_output = new DigitalOut((PinName)desc->number);
+		if (!gpio_output)
+			return -ENOMEM;
+
+		gpio_desc_extra->gpio_pin = (struct mbed_gpio_desc *)gpio_output;
+		gpio_desc_extra->direction = GPIO_OUT;
+
+		/* Set the GPIO value */
+		if (gpio_set_value(desc, value) != SUCCESS) {
```

Comment:
Would it be better to use `gpio_output->write(value)`? `gpio_set_value` seems a bit unclear what it will do.

Reply:
This is good point. We can use write() function directly.